src/resonator_fits.jl

```diff
@@ -0,0 +1,300 @@
+    f0: Resonator center frequency.
+    Qi: Resonator internal quality factor.
+    Qc: Resonator coupling quality factor.
+    ϕ: Impedance mismatch angle.
+    τ: System phase delay.
+    α: System gain angle.
+    A: System gain amplitude.
+  """
+  f0
+  Qi
+  Qc
+  ϕ
+  τ
+  α
+  A
+end
+
+#Fitting to the resonance circle of a quarter-wave resonator
+function fit_resonance_circle{T <: AbstractFloat}(freq::Array{T, 1}, data::Array{Complex{T}, 1})
+  """
+  Fits complex-valued data
+
+  Args:
+    freq: Frequency data.
+    data: Complex S21 data.
+  Returns:
+    Result: The result of the fit.
+  """
+  @assert length(freq) == length(data) "Frequency and Data vectors must have same length!"
+  @assert length(data) > 20 "Too few points."
+
+  τ, α, A = calibrate_resonance_circle(freq, data)
+  scaled_data = apply_calibration(τ, α, A, freq, data)
+
+  R, xc, yc = fit_circle(real(scaled_data), imag(scaled_data))
+  ϕ = -asin(yc / R)
+  td = (real(scaled_data) - xc) + 1im*(imag(scaled_data) - yc)
+  f0, Qfit, _ = fit_phase(freq, td)
+  Qc = Qfit / (2. * R * exp(-1im * ϕ))
+  Qi = 1./ (1./Qfit - real(1./Qc))
+  Qc = abs(Qc)
+
+  return CircleFitResult(f0, Qi, Qc, ϕ, -τ, -α, A)
+end
+
+function calibrate_resonance_circle(freq, data)
+  """Calibrates out cable delay and overall system gain to translate resonance
+  circle to "canonical" position for fit.
+
+  Args:
+    freq: Frequency data.
+    data: Complex S21 data.
+  Returns:
+    τ: Fitted cable delay.
+    α: Overall system gain angle.
+    A: Overall system gain amplitude.
+  """
+  #Fit to cable delay.
+  τ = fit_delay(freq, data)
+  Sp = exp(-1im * 2. * π * freq * τ) .* data
+  #Get best-fit circle and translate to origin.
+  R, xc, yc = fit_circle(real(Sp), imag(Sp))
+  Strans = (real(Sp) - xc) + 1im * (imag(Sp) - yc)
+  #Calculate invariant ω → ∞ point
+  _, _, θ = fit_phase(freq, Strans)
+  P = xc + R * cos(θ + π) + 1im * (yc + R * sin(θ + π))
+  A = abs(P)
+  α = angle(P)
+  return τ, α, A
+end
+
+function apply_calibration(τ, α, A, freq, data)
+  """Apply the calibration to the resonance data to move it to the
+  canonical position.
+
+  Args:
+    τ: Fitted cable delay.
+    α: Overall system gain angle.
+    A: Overall system gain amplitude.
+    freq: Frequency data.
+    data: Complex S21 data.
+  Returns:
+    scaled_data: Data shifted to the canonical position for a circle fit.
+  """
+  data = data .* exp(-1im * 2 * π * τ * freq)
+  rot = [cos(-α) -sin(-α); sin(-α) cos(-α)]
+  scaled_data = zeros(Complex128, size(data))
+  for j = 1:length(data)
+    v = rot * [real(data[j]); imag(data[j])]
+    scaled_data[j] = (v[1] + 1im * v[2]) / A
+  end
+  return scaled_data
+end
+
+function fit_phase(freq, data)
+  """Fit phase of resonance.
+  Args:
+    freq: Frequency data.
+    data: Complex S21 data.
+  Returns:
+    f0: Center frequency.
+    Q: Quality factor.
+    Θ0: Offset phase angle.
+  """
+  model(x, p) = p[1] + 2. * slope * atan(2 * p[2] *(1. - x / p[3]))
```

Comment:
Last time I tried writing something similar, I needed to use `atan2` to get a function that would have a range of 2π. Is that not necessary here?